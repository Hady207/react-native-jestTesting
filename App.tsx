/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * Generated with the TypeScript template
 * https://github.com/react-native-community/react-native-template-typescript
 *
 * @format
 */

import React, {type PropsWithChildren, useState} from 'react';
import {
  SafeAreaView,
  ScrollView,
  StatusBar,
  StyleSheet,
  Text,
  useColorScheme,
  View,
  Button,
  FlatList,
  TextInput,
} from 'react-native';

import {Colors, Header} from 'react-native/Libraries/NewAppScreen';

export const Section: React.FC<
  PropsWithChildren<{
    title: string;
  }>
> = ({title}) => {
  const isDarkMode = useColorScheme() === 'dark';
  const [count, setCount] = useState<number>(1);
  return (
    <View style={styles.sectionContainer}>
      <View
        style={{
          backgroundColor: 'red',
          flexDirection: 'row',
          alignItems: 'center',
        }}>
        <Text
          style={[
            styles.sectionTitle,
            {
              color: isDarkMode ? Colors.white : Colors.black,
            },
          ]}>
          {title}
        </Text>
        <Text
          style={[
            styles.sectionDescription,
            {
              color: isDarkMode ? Colors.light : Colors.dark,
            },
          ]}
          testID="counterId">
          {count}
        </Text>
      </View>

      <Button
        color={'green'}
        title="+"
        testID="plus-button"
        onPress={() => setCount(sw => sw + 1)}
      />
      <Button
        color={'red'}
        title="-"
        testID="minus-button"
        onPress={() => setCount(sw => sw - 1)}
      />
    </View>
  );
};

export const FlatSection = ({onEndReached}: {onEndReached?: () => void}) => {
  return (
    <FlatList
      contentContainerStyle={{
        flexGrow: 1,
        backgroundColor: 'yellow',
        height: 50,
      }}
      data={[{}, {}, {}, {}, {}, {}, {}, {}, {}, {}]}
      onEndReached={onEndReached}
      horizontal
      onEndReachedThreshold={0.2}
      testID="flat-list"
      renderItem={() => (
        <View
          style={{
            backgroundColor: 'red',
            height: 10,
            width: 50,
            marginRight: 10,
          }}
        />
      )}
    />
  );
};

export const FormSection = () => {
  const [name, setName] = useState('');
  const [show, setShow] = useState(false);
  return (
    <View>
      <TextInput placeholder="name" value={name} onChangeText={setName} />
      <Button
        title="print"
        testID="print-button"
        onPress={() =>
          setTimeout(() => setShow(true), Math.floor(Math.random() * 200))
        }
      />
      {show && <Text testID="printed-username">{name}</Text>}
    </View>
  );
};

const App = () => {
  const isDarkMode = useColorScheme() === 'dark';

  const backgroundStyle = {
    backgroundColor: isDarkMode ? Colors.darker : Colors.lighter,
  };

  return (
    <SafeAreaView style={backgroundStyle}>
      <StatusBar barStyle={isDarkMode ? 'light-content' : 'dark-content'} />

      <ScrollView
        contentInsetAdjustmentBehavior="automatic"
        style={backgroundStyle}>
        <Header />
        <View
          style={{
            backgroundColor: isDarkMode ? Colors.black : Colors.white,
          }}>
          {/* <LearnMoreLinks /> */}
          <Section title="Count" />
          <FlatSection onEndReached={() => console.log('finished')} />

          <FormSection />
        </View>
      </ScrollView>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  sectionContainer: {
    marginTop: 32,
    paddingHorizontal: 24,
  },
  sectionTitle: {
    fontSize: 24,
    fontWeight: '600',
  },
  sectionDescription: {
    marginLeft: 12,
    fontSize: 24,
    fontWeight: '600',
  },
  highlight: {
    fontWeight: '700',
  },
});

export default App;
