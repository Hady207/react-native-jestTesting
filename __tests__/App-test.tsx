/**
 * @format
 */
// import '@testing-library/jest-native';
import '@testing-library/jest-native/extend-expect';
import 'react-native';
import React from 'react';
import App, {Section, FlatSection, FormSection} from '../App';
import {
  render,
  screen,
  fireEvent,
  waitFor,
} from '@testing-library/react-native';

// Note: test renderer must be required after react-native.
import '@testing-library/jest-native/extend-expect';
import renderer from 'react-test-renderer';

test('renders correctly', () => {
  renderer.create(<App />);
});

test('render text in section', () => {
  render(<Section title="count " />);
  const PlusButton = screen.getByText('+');
  const MinusButton = screen.getByText('-');
  const Counter = screen.getByText('1');

  fireEvent.press(PlusButton);
  expect(Counter).toHaveTextContent('2');
  fireEvent.press(MinusButton);
  expect(Counter).toHaveTextContent('1');
  fireEvent.press(MinusButton);
  expect(Counter).toHaveTextContent('0');
});

test('test scrolling flatsection', () => {
  const onEndReached = jest.fn();
  render(<FlatSection onEndReached={onEndReached} />);
  const eventData = {
    nativeEvent: {
      contentOffset: {
        x: 300,
      },
      contentSize: {
        // Dimensions of the scrollable content
        height: 10,
        width: 50,
      },
      layoutMeasurement: {
        // Dimensions of the device
        height: 100,
        width: 100,
      },
    },
  };
  fireEvent.scroll(screen.getByTestId('flat-list'), eventData);
  expect(onEndReached).toHaveBeenCalled();
});

test('Test Form in react native', async () => {
  const {toJSON} = render(<FormSection />);
  const programmer = 'Hadi Maher';
  const textInput = screen.getByPlaceholderText('name');
  const button = screen.getByTestId('print-button');

  fireEvent.changeText(textInput, programmer);
  fireEvent.press(button);

  await waitFor(() =>
    expect(screen.getByTestId('printed-username')).toBeTruthy(),
  );

  expect(screen.getByTestId('printed-username').props.children).toBe(
    programmer,
  );

  expect(toJSON()).toMatchSnapshot();
});
