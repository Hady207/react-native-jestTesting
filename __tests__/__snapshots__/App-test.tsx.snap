// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Form in react native 1`] = `
<View>
  <TextInput
    onChangeText={[Function]}
    placeholder="name"
    value="Hadi Maher"
  />
  <View
    accessibilityRole="button"
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="print-button"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        print
      </Text>
    </View>
  </View>
  <Text
    testID="printed-username"
  >
    Hadi Maher
  </Text>
</View>
`;
